### DISCUSSION TITLE
Default ResponseHandler for Server Sent Event (SSE)

### DISCUSSION THREAD

**Discussion by galileilei** [earthboundkid/requests#120] [Ideas] [answered]
If I want to use`requests` on http endpoint that returns Server Sent Event (SSE) (for example on OpenAI chat completion streaming endpoint), a workaround is implementing a SSE parser ([specification here](https://html.spec.whatwg.org/multipage/server-sent-events.html#parsing-an-event-stream)) on `bufio.Scanner`. 

This functionality (SSE parser) may be outside of the scope of `requests`, it is frequent enough that I think a `requests/sse` sub-package would be a great addition.

**Comment by earthboundkid** [comment]
That's an interesting idea. I'm not sure how good of a fit requests would be since the idea in requests is that the requests is that each request is handled in a single function call. For SSE, I imagine you would want to open a loop of some sort. What kind of API were you imagining?

**Reply by galileilei** [reply]
Something like this:

```go
es := sse.NewEventSource(ctx) // (1)
err := requests.URL("").BodyJSON(&body).Handle(es.Recv).Fetch(ctx) // (2)
for event in es.All() { // (3)
     // process the event
}
```
A few thoughts:
1. Existing packages I found are [ssereader](https://github.com/mojocn/sseread) and [go-sse](https://github.com/tmaxmax/go-sse). Have not looked into integrating them with requests package though.
2. `es.Recv` does not block and return immediately. Here I use the generic `Handle()` function; having a dedicated `ToSSE()` can be an option too.
3. Use iterator style, instead of callback style in [Javascript EventSource example](https://developer.mozilla.org/en-US/docs/Web/API/EventSource#examples) or [go-sse example](https://github.com/tmaxmax/go-sse?tab=readme-ov-file#subscribing-to-events). The iterator can either a method with pointer receiver `es`, or a function on `es`.

**Reply by earthboundkid** [reply]
Maybe there could be a magic handler that is blest to leave the connection open at the end of the request. Something like

```go
var rc io.ReadCloser
err := requests.
    URL("http://example.com/sse").
    BodyJSON(&body).
    CheckContentType("text/event-stream").
    Handle(requests.KeepOpen(&rc)).
    Fetch(ctx)
if err !=nil { /**/ }
defer rc.Close()
for event, err := sse.ReadEvents(rc) {
    // ...
}
```

**Reply by earthboundkid** [reply, marked as answer]
The go-sse client already does a lot of the validation and backoff stuff you might want to use requests for. I think the solution might be to just use it with requests as a simple request builder. From their example:

```go
client := sse.Client{
    Backoff: sse.Backoff{
        MaxRetries: -1,
    },
    // other settings...
}

body := whatever()
req, err := requests.URL("https://api.openai.com/...").BodyJSON(&body).Request(ctx)
if err != nil { /**/ }
conn := client.NewConnection(req)

conn.SubscribeMessages(/* callback */)

if err := conn.Connect(); !errors.Is(err, io.EOF) {
    // handle error
}
```

**Reply by galileilei** [reply]
Did not know there is `Request()` API. will try this out!
